### PR TITLE
GH-40749: [Python][Packaging] Strip unnecessary symbols when building wheels

### PR DESCRIPTION
**Pull request by raulcd**
### Rationale for this change

Removing unnecessary symbols for wheels will allow us to reduce the size of the wheels considerably.

### What changes are included in this PR?

Running `strip --strip-debug` on Linux wheels for all *.so files.

### Are these changes tested?

Yes, via Archery.

### Are there any user-facing changes?

No
* GitHub Issue: #40749

### CODE REVIEW

**Comment by raulcd** [comment]
@github-actions crossbow submit wheel-manylinux-*

**Comment by raulcd** [comment]
This change reduces from [131MB locally unzipped wheel](https://github.com/ursacomputing/crossbow/actions/runs/9397680127) from the previous nightly build to [101MB locally unzipped wheel](https://github.com/ursacomputing/crossbow/actions/runs/9418506664) the wheel on wheel-manylinux-2-28-cp310-amd64.

**Comment by raulcd** [comment]
@jorisvandenbossche @pitrou would you prefer to:
- [1] Go with the current change just using `--strip` on `auditwheel repair`
- [2] Run `strip --strip-debug` manually
- [3] Wait until https://github.com/pypa/auditwheel/issues/491 is potentially done and released

**Comment by raulcd** [comment]
@github-actions crossbow submit wheel-manylinux-*

**Comment by raulcd** [comment]
@github-actions crossbow submit wheel-manylinux-*

**Comment by raulcd** [comment]
Running `strip --strip-debug` manually seems to reduce a locally unzipped wheel from 131MB to 124MB

**Comment by raulcd** [comment]
`strip -S $filename` doesn't seem to have any effect on macOS, the size of the wheels and the unpacked wheel is exactly the same with and without executing the command

File: ci/scripts/python_wheel_manylinux_build.sh [hunk 1]
```diff
@@ -160,6 +160,26 @@ export CMAKE_PREFIX_PATH=/tmp/arrow-dist
...
+wheel_name=$(ls pyarrow-*.whl)
+# Unzip and remove old wheel
+unzip $wheel_name
+rm $wheel_name
+for filename in $(ls pyarrow/*.so pyarrow/*.so.*); do
+    echo "Stripping debug symbols from: $filename";
+    strip --strip-debug $filename
+done
+# Zip wheel again after stripping symbols
+zip -r $wheel_name .
```
Comment (pitrou) [inline ci/scripts/python_wheel_manylinux_build.sh:177]: Does this reuse the same compression settings? (is a wheel compressed at all?)

Comment (raulcd) [reply ci/scripts/python_wheel_manylinux_build.sh:177]: Both wheels are compressed, a size comparison shows:
Initial one from CI from main:
```
-rw-r--r--  1 raulcd raulcd  40M jun  6 08:46 pyarrow-17.0.0.dev260-cp310-cp310-manylinux_2_28_x86_64.whl
```
and new one generated from this PR
```
-rw-r--r--  1 raulcd raulcd  38M jun  7 17:35 pyarrow-17.0.0.dev275-cp310-cp310-manylinux_2_28_x86_64.whl
```

The compression type for the files is exactly the same and all file sizes are exactly the same, I've tested with the following snippet:
```
import zipfile

with zipfile.ZipFile('pyarrow-17.0.0.dev260-cp38-cp38-manylinux_2_28_x86_64.whl', 'r') as old_wheel:
    with zipfile.ZipFile('pyarrow-17.0.0.dev275-cp38-cp38-manylinux_2_28_x86_64.whl', 'r') as new_wheel:
        for info_old, info_new in zip(sorted(old_wheel.infolist(), key=lambda x: x.filename), sorted(new_wheel.infolist(), key=lambda x: x.filename)):
            if "dev260" not in info_old.filename:
                assert info_old.filename == info_new.filename, info_old.filename + info_new.filename
            assert info_old.compress_type == info_new.compress_type
            if "RECORD" in info_old.filename:
                continue
            if not (info_old.filename.endswith('.so') or info_old.filename.endswith('.so.1700')):
                assert info_old.compress_size == info_new.compress_size, f"old: {info_old.filename, info_old.compress_size}, new: {info_new.filename, info_new.compress_size}"
            print(f'filename: {info_old.filename} ')
            print(f'compress type: {info_old.compress_type}')
            print(f'compress size: {info_new.compress_size}')
```
I am not showing the full output but this a an example of what I get:
```
...
filename: pyarrow/vendored/docscrape.py 
compress type: 8
compress size: 6080
filename: pyarrow/vendored/version.py 
compress type: 8
compress size: 3894
filename: scripts/ 
compress type: 0
compress size: 0
filename: scripts/test_imports.py 
compress type: 8
compress size: 496
filename: scripts/test_leak.py 
compress type: 8
compress size: 1422

```

I am not an expert on zip compression but all tests for the wheels are successful on the CI jobs and my validations seem to show the exact same level of compression and same size for all files in both wheels.
Any other test you think I should do or how I could check?

Comment (pitrou) [reply ci/scripts/python_wheel_manylinux_build.sh:177]: You're right, this does seem to work and the fact that the wheels are shrunk seems to indicate it's ok. Thank you!

File: ci/scripts/python_wheel_manylinux_build.sh [hunk 2]
```diff
@@ -160,6 +160,26 @@ export CMAKE_PREFIX_PATH=/tmp/arrow-dist
 pushd /arrow/python
 python setup.py bdist_wheel
 
+echo "=== Strip symbols from wheel ==="
+mkdir dist/temp-fix-wheel
```
Comment (pitrou) [inline ci/scripts/python_wheel_manylinux_build.sh:164]: Let's avoid erroring out if the directory already exists?
```suggestion
mkdir -p dist/temp-fix-wheel
```

**Review by pitrou — APPROVED** [review]
+1 except for a nit

**Comment by raulcd** [comment]
:+1: merging

**Pull request merged**